File: C++/Problems/Problems/problem_841.cpp
#include "problem_841.h"

#include <stack>

bool problem_841::test() {
    std::vector<std::vector<int>> rooms { {1}, {2}, {3}, {} };

    bool expected = true;

    auto result = canVisitAllRooms(rooms);

    return expected == result;
}

bool problem_841::canVisitAllRooms(std::vector<std::vector<int>> &rooms) {
    int n = rooms.size();
    int covered = 0;
    std::vector visited(n, false);
    std::stack<int> toVisit;
    toVisit.push(0);
    while (!toVisit.empty())
    {
        auto next  = toVisit.top();
        toVisit.pop();
        if (visited[next])
            continue;
        covered++;
        visited[next] = true;
        for (auto key : rooms[next])
        {
            if (!visited[key])
                toVisit.push(key);
        }
    }

    return covered == n;
}

File: Python/Problems/Problems/problem841.py
import heapq
import math
import sys

from utils import print1DArray, print2DArray
from typing import List, Set
from collections import Counter

from .problem import Problem


class Problem841(Problem):
    def test(self) -> bool:
        rooms = [[1], [2], [3], []]

        expected = True

        result = self.canVisitAllRooms(rooms)

        return expected == result

    def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
        n = len(rooms)
        covered = 0
        visited = [False] * n
        toVisit = []
        toVisit.append(0)
        while toVisit:
            next = toVisit.pop()
            if visited[next]:
                continue
            covered += 1
            visited[next] = True
            for key in rooms[next]:
                if not visited[key]:
                    toVisit.append(key)

        return covered == n